### PR TITLE
Features/rio 351

### PR DESCRIPTION
**Pull request by jamiequish**

### CODE REVIEW

File: Source/Rio.API/Controllers/ParcelLedgerController.cs [hunk 1]
```diff
@@ -33,11 +34,22 @@ public IActionResult New([FromBody] ParcelLedgerCreateDto parcelLedgerCreateDto)
             var parcelDto = Parcel.GetByParcelNumberAsDto(_dbContext, parcelLedgerCreateDto.ParcelNumbers[0]); 
             if (parcelDto == null)
             {
-                ModelState.AddModelError("ParcelNumber", $"{parcelLedgerCreateDto.ParcelNumbers[0]} is not a valid Parcel APN.");
-                return BadRequest(ModelState);
+                if (parcelLedgerCreateDto.ParcelNumbers[0] == null)
```
Comment (raynardlee) [inline Source/Rio.API/Controllers/ParcelLedgerController.cs:37]: are you always passing an array of at least one here?  I would think you would need to check if you have any ParcelNumbers first

File: Source/Rio.API/Controllers/ParcelLedgerController.cs [hunk 2]
```diff
@@ -33,11 +34,22 @@ public IActionResult New([FromBody] ParcelLedgerCreateDto parcelLedgerCreateDto)
...
+                }
+            }
+
+            if (parcelLedgerCreateDto.TransactionTypeID == (int) TransactionTypeEnum.Supply && parcelLedgerCreateDto.WaterTypeID == null)
+            {
+                ModelState.AddModelError("SupplyType", "The Supply Type field is required for transactions adjusting water supply.");
             }
 
-            ValidateEffectiveDate(parcelLedgerCreateDto.EffectiveDate);
+            ValidateEffectiveDate(parcelLedgerCreateDto.EffectiveDate.Value, parcelLedgerCreateDto.EffectiveDateString);
```
Comment (raynardlee) [inline Source/Rio.API/Controllers/ParcelLedgerController.cs:52]: why is there a string and a date version?  why are we using both?

File: Source/Rio.API/Models/ParcelLedgerCreateCSVUploadDto.cs [hunk 3]
```diff
@@ -10,6 +10,7 @@ public class ParcelLedgerCreateCSVUploadDto
         public IFormFile UploadedFile { get; set; }
         [Required]
         public DateTime EffectiveDate { get; set; }
+        public String EffectiveDateString { get; set; }
```
Comment (raynardlee) [inline Source/Rio.API/Models/ParcelLedgerCreateCSVUploadDto.cs:13]: lower case string
also I need to understand why you have both a DateTime version and a string version

File: Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs [hunk 4]
```diff
@@ -2,19 +2,23 @@
 using System.Collections.Generic;
 using System.ComponentModel.DataAnnotations;
 
-namespace Rio.Models.DataTransferObjects.ParcelLedgerCreateDto
+namespace Rio.Models.DataTransferObjects.ParcelAllocation
```
Comment (raynardlee) [inline Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs:5]: Why do we have an Allocation reference again?

File: Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs [hunk 5]
```diff
@@ -2,19 +2,23 @@
...
     public class ParcelLedgerCreateDto
     {
-        [Required]
+        [Required] 
         public List<string> ParcelNumbers { get; set; }
+        [Display(Name = "Effective Date")]
         [Required]
-        public DateTime EffectiveDate { get; set; }
+        public DateTime? EffectiveDate { get; set; }
+        public string EffectiveDateString { get; set; }
```
Comment (raynardlee) [inline Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs:14]: Again, why EffectiveDate and EffectiveDateString

File: Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts [hunk 6]
```diff
@@ -75,22 +74,33 @@ export class ParcelLedgerCreateFromSpreadsheetComponent implements OnInit {
...
+
+    if (!this.validateEffectiveDateAndWaterTypeID()) {
+      this.isLoadingSubmit = false;
+      window.scroll(0,0);
+      return;
+    }
 
     const _datePipe = this.datePipe;
-    const effectiveDateAsString = _datePipe.transform(this.effectiveDate, 'MM/dd/yyyy');
+    const effectiveDateString = (<HTMLInputElement> document.getElementById("effective-date")).value;
```
Comment (raynardlee) [inline Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts:101]: why are we using getElementById?

File: Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts [hunk 7]
```diff
@@ -75,22 +74,33 @@ export class ParcelLedgerCreateFromSpreadsheetComponent implements OnInit {
     this.document.getElementById("CSV-upload").click();
   }
 
-  private clearErrorAlerts() {
-    if (!this.alertsCountOnLoad) {
-      this.alertsCountOnLoad = this.alertService.getAlerts().length;
+  private validateEffectiveDateAndWaterTypeID() {
+    var formValid = true;
+    if (!this.effectiveDate) {
```
Comment (raynardlee) [inline Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts:79]: Why isn't this handled by the server side validators?

File: Source/Rio.Web/src/app/services/parcel-ledger.service.ts [hunk 8]
```diff
@@ -21,10 +21,11 @@ export class ParcelLedgerService {
     return this.apiService.postToApi(route, model);
   }
 
-  newCSVUploadTransaction(uploadedFile: any, effectiveDateAsString: string, waterTypeID: number): Observable<any[]> {
+  newCSVUploadTransaction(uploadedFile: any, effectiveDateString: string, waterTypeID: number): Observable<any[]> {
     let formData = new FormData();
     formData.append("UploadedFile", uploadedFile);
-    formData.append("EffectiveDate", effectiveDateAsString);
+    formData.append("EffectiveDate", effectiveDateString);
```
Comment (raynardlee) [inline Source/Rio.Web/src/app/services/parcel-ledger.service.ts:27]: again, why are we doing this?

File: Source/Rio.API/Controllers/ParcelLedgerController.cs (hunk 1 above)
Comment (jamiequish) [reply Source/Rio.API/Controllers/ParcelLedgerController.cs:37]: The form just sends the inputted parcel number as a list so we can reuse the dto, so we can assume there will be one value but that the value might null.

File: Source/Rio.API/Controllers/ParcelLedgerController.cs (hunk 2 above)
Comment (jamiequish) [reply Source/Rio.API/Controllers/ParcelLedgerController.cs:52]: I need to validate the user-inputted string to make sure it's in the right YYYY-MM-DD format, but I can just pass the string and then parse it as a date after it's validated.

File: Source/Rio.API/Controllers/ParcelLedgerController.cs (hunk 1 above)
Comment (jamiequish) [reply Source/Rio.API/Controllers/ParcelLedgerController.cs]: Function reworked per the recommendations made earlier

File: Source/Rio.API/Controllers/ParcelLedgerController.cs (hunk 2 above)
Comment (jamiequish) [reply Source/Rio.API/Controllers/ParcelLedgerController.cs]: EffectiveDateString field removed, EffectiveDate is now a string and format is validated via RegularExpressionAttribute before parsing as DateTime.

File: Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts (hunk 7 above)
Comment (jamiequish) [reply Source/Rio.Web/src/app/pages/parcel-ledger-create-from-spreadsheet/parcel-ledger-create-from-spreadsheet.component.ts]: Validation moved to API

File: Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs (hunk 4 above)
Comment (jamiequish) [reply Source/Rio.Models/DataTransferObjects/ParcelAllocation/ParcelLedgerCreateDto.cs]: Folder renamed

**Pull request merged**